addons/udes_stock/tests/test_picking_batch.py

```diff
@@ -1087,49 +1087,110 @@ class TestBatchGetNextTask(common.BaseUDES):
 
         # In the same way, we should now get '10'
         self.assertEqual(task['package_id']['name'], '10')
 
+    def test02_skip_products(self):
+        """Test that get_next_task respects skipped products"""
+        # Assert that the specified product is not in the next task
+        for product_id in [self.apple.id, self.banana.id]:
+            task = self.batch.get_next_task(skipped_product_ids=[product_id])
+            task_quants = task['package_id']['quant_ids']
+            task_products = [quant["product_id"]["id"] for quant in task_quants]
+            self.assertNotIn(product_id, task_products)
+
+    def test03_skip_products_return(self):
+        """Test that get_next_task can return to skipped products
+        if u_return_to_skipped is True.
+        """
+        # First assert if we skip everything with default config, the returned task is empty
+        task = self.batch.get_next_task(skipped_product_ids=[self.apple.id, self.banana.id])
+        self.assertFalse(task["move_line_ids"])
+        self.assertFalse(task.get("package_id"))
+        # Change config to allow returning to skipped tasks
+        self.picking_type_pick.u_return_to_skipped = True
+        # First returned task should include the first skipped product
+        for product_ids in [[self.apple.id, self.banana.id], [self.banana.id, self.apple.id]]:
+            task = self.batch.get_next_task(skipped_product_ids=product_ids)
+            task_quants = task['package_id']['quant_ids']
+            task_products = [quant["product_id"]["id"] for quant in task_quants]
+            self.assertIn(product_ids[0], task_products)
+            self.assertNotIn(product_ids[1], task_products)
+
+    def test04_skip_move_lines(self):
+        """Test that get_next_task respects skipped products"""
+        move_line_ids = self.batch.picking_ids.mapped("move_line_ids.id")
+        # Assert that the specified move_line_id is not in the next task
+        for move_line_id in move_line_ids:
+            task = self.batch.get_next_task(skipped_move_line_ids=[move_line_id])
+            self.assertNotIn(move_line_id, task['move_line_ids'])
+
+    def test05_skip_move_lines_return(self):
+        """Test that get_next_task can return to skipped products
+        if u_return_to_skipped is True.
+        """
+        move_line_ids = self.batch.picking_ids.mapped("move_line_ids.id")
+        # First assert if we skip everything with default config, the returned task is empty
+        task = self.batch.get_next_task(skipped_move_line_ids=move_line_ids)
+        self.assertFalse(task["move_line_ids"])
+        self.assertFalse(task.get("package_id"))
+        # Change config to allow returning to skipped tasks
+        self.picking_type_pick.u_return_to_skipped = True
+        # First returned task should include the first skipped move_line_id
+        for skipped_move_line_ids in [move_line_ids, [ml for ml in reversed(move_line_ids)]]:
```

Comment:
Q: Why is move_line_ids reversed?

Reply:
So we re-try the below test with the move lines reversed to check that the returned task changes to match the first move line given to it. So if our move lines are `[3,4]`, on the first iteration of the loop the task we expect to see should include moveline `3`. On the second iteration when these are reversed we expect to get the task related to moveline `4` returned. This is mainly done as a sanity check to make sure it will pick out the first move-line preferentially